### ISSUE TITLE
listsinceblock and listtransactions don't show incoming unconfirmed transactions

### ISSUE THREAD

**Issue by santins77** [gridcoin-community/Gridcoin-Research#1157]
The list of transactions returned by listsinceblock and listtransactions don't show the incoming transactions with 0 confirmations. They are only shown when they reach one confirmation. This rpc api behavior is different comparing to bitcoin and some other bitcoin family coins

**Comment by iFoggz** [comment]
the `listtransactions` functionality is not much different then bitcoins actually . if the transaction has 0 confirms it does show for me in `listtransactions` with 0 confirms (sent to myself). default it shows only 10 transactions. there are some issues regarding `listsinceblock` in the main release that is addressed in development slated for release in near future.i'll verify there were no changes to `listtransactions` since release

only time no tx would be shown is if it conflicted `-1`

also note that if someone sends u a tx u wont see it till 1 confirm anyway as the block u receive that has the tx is considered first confirm

**Comment by santins77** [comment]
I'm having the issue when the transaction is sent from some external address to my wallet. When I sent money from my wallet to another address the commands seem to work OK. That is why I mention in incoming transactions.

**Comment by iFoggz** [comment]
so your sending from a wallet to another wallet then? or are you sending between two addresses within same wallet?

**Comment by iFoggz** [comment]
any txs received from the network from two different wallets wont show till 1 confirm as it sits in the memory pool until a staked block carries the transaction

**Comment by santins77** [comment]
Do you have in mind to change that behavior? In bitcoin, lite, and others the show this transactions in those cases

**Comment by iFoggz** [comment]
that would likely occur when we do a rebase to add the functionality of showing transactions sitting in the memory pool.

**Comment by iFoggz** [comment]
we will be re-basing to update the long overdue code in the files

**Comment by santins77** [comment]
OK. Nice! Thanks for the info

**Comment by iFoggz** [comment]
`getrawmempool` rpc command will display all txs in the memory pool waiting to be staked in a block. you can still `gettransaction` a tx if it is in the memorypool

**Comment by iFoggz** [comment]
this behavior still present?

**Comment by iFoggz** [comment]
So I don't know if I consider this an overall problem still:

listtransactions:
If sent to myself: will show with 0 confirmations
if received from someone else: will show after 1 confirmation
immature stakes will show immediately as its technically confirmed 1 time unless the network disagrees and a reorg occurs at a later time in which they shouldn't be showing anymore as they are -1 conflicted and they are ignored.

I don't agree with a unconfirmed transaction showing up unless at least 1 confirm but that my opinion. It is in the memory pool awaiting to be accepted into the network. 

listsinceblock:
this is block based so if a transaction is not in a block then I see no reason for the wallet to show it.

It is just my opinion that a transaction with 0 confirms should not show in listtransaction nor GUI as well. I make no plan(s) to touch this issue myself. I think it would be more confusing for the non technical user to see a transaction appear and only to disappear after the next block because it was rejected by the network for any number of reasons. 

If no one else agrees they could take this one on.

**Comment by joewinkler** [comment]
I think it is more confusing, a transaction you've sent by yourself not being in the transaction list. You may be confused, whether you really did this transaction. If it is clearly visible there are no confirmations yet, I cannot see any problem in showing transactions without confirmation.

**Comment by iFoggz** [comment]
a transaction being sent by yourself to yourself will show as well as a transaction you sent will show as well.

What i was saying is I don't think it should show a incoming transaction to you from someone else that hasnt had atleast 1 confirm. 

The original poster wanted to see 0 confirm incoming transactions from other people to him.

**Comment by jamescowens** [comment]
I agree Paul. We should close this issue.

**Comment by iFoggz** [comment]
I agree, any input from @cyrossignol  and @div72  on this?

**Comment by div72** [comment]
I disagree.

Imagine a situation where I am a vendor and you are a long time customer buying a low value item at a market. I would rather not wait ~96 seconds before the tx gets in to a block and just accept the payment when it hits my mempool since I trust you and there is low risk and low loss.

Irregardless of what I think though, we should follow upstream in order to maintain compatibility with scripts designed for it.

**Comment by jamescowens** [comment]
@div72 has a good point.

**Comment by iFoggz** [comment]
@div72 I do agree to that point however we still not there so this will likely be pushed off repeatedly since all the files are changed as well as locations of things upstream.

**Comment by div72** [comment]
@iFoggz Alright, let's do the RPC port then. I'll start that tomorrow.

**Comment by jamescowens** [comment]
We need to be super careful during the port that we do not unwittingly break Gridcoin specific stuff. The port is going to be irritating I think... :)

**Comment by iFoggz** [comment]
it wouldnt just be rpc part to be ported but all the src/wallet/rpc src/wallet parts. they broke stuff up and moved alot around

to me this isnt an issue but rather a goal apart of the portign of all the bitcoin code is more what i was getting at.

**Comment by PrestackI** [comment]
# Issue #1157: Show Incoming Unconfirmed Transactions in RPC Commands
## Complete Analysis & Implementation Strategy

**Issue URL**: https://github.com/gridcoin-community/Gridcoin-Research/issues/1157  
**Status**: Open (Milestone: Opossum)  
**Created**: June 8, 2018  
**Last Updated**: August 5, 2025  
**Priority**: High (P1 - Mempool Commands per RPC Porting Analysis)

---

## Table of Contents

1. [Executive Summary](#executive-summary)
2. [Problem Description](#problem-description)
3. [Root Cause Analysis](#root-cause-analysis)
4. [Architecture Comparison](#architecture-comparison)
5. [Proposed Solution](#proposed-solution)
6. [Implementation Strategy](#implementation-strategy)
7. [Timeline & Resources](#timeline--resources)
8. [Success Criteria](#success-criteria)
9. [Related Issues](#related-issues)

---

## Executive Summary

### The Problem
The RPC commands `listtransactions` and `listsinceblock` do not show incoming transactions with 0 confirmations. These transactions only appear after receiving their first confirmation (being included in a block). This behavior differs from Bitcoin Core and other Bitcoin-family cryptocurrencies, breaking compatibility with scripts and tools designed for Bitcoin's RPC API.

### Root Cause
The issue is **NOT** in the RPC command implementations themselves, but rather in how the wallet tracks transactions. Gridcoin's wallet currently only adds incoming transactions to `mapWallet` (the wallet's transaction map) after they've been confirmed in a block. Bitcoin Core adds incoming transactions to `mapWallet` immediately when they enter the mempool.

### Solution Approach
Port Bitcoin Core's `SyncTransaction` architecture to Gridcoin. This involves:
1. Creating explicit transaction state types
2. Implementing a unified transaction synchronization mechanism
3. Integrating with validation layer callbacks
4. Updating all wallet transaction handling

### Implementation Scope
- **Effort**: 10-12 weeks (2.5-3 months)
- **Complexity**: High (architectural refactor)
- **Risk**: Medium (comprehensive testing mitigates)
- **Benefit**: High (foundation for future improvements)

---

## Problem Description

### User Report

From issue #1157 (santins77):
> "The list of transactions returned by listsinceblock and listtransactions don't show the incoming transactions with 0 confirmations. They are only shown when they reach one confirmation. This rpc api behavior is different comparing to bitcoin and some other bitcoin family coins"

### Affected RPC Commands

1. **`listtransactions`** - Lists recent transactions
   - **Expected**: Show all transactions including unconfirmed (0 confirmations)
   - **Actual**: Only shows incoming transactions after 1+ confirmations

2. **`listsinceblock`** - Lists transactions since a specific block
   - **Expected**: Show all transactions including unconfirmed
   - **Actual**: Only shows incoming transactions after 1+ confirmations

### Current Behavior vs Expected Behavior

| Transaction Type | Current Behavior | Expected Behavior | Bitcoin Behavior |
|------------------|------------------|-------------------|------------------|
| **Outgoing (sent from wallet)** | ✅ Shows immediately with 0 confirmations | ✅ Shows immediately | ✅ Shows immediately |
| **Incoming (received from external)** | ❌ Only shows after 1+ confirmations | ✅ Shows immediately with 0 confirmations | ✅ Shows immediately |

### Why This Matters

1. **API Compatibility**: Tools and scripts written for Bitcoin expect unconfirmed transactions to appear
2. **User Experience**: Users can't see incoming payments until confirmed (~90 seconds in Gridcoin)
3. **Payment Workflows**: Merchants may want to display unconfirmed payments with appropriate warnings
4. **Development**: Makes Gridcoin's RPC API inconsistent with Bitcoin Core

### Community Consensus

From issue comments:
- **div72** (Member): "Irregardless of what I think though, we should follow upstream in order to maintain compatibility with scripts designed for it."
- **jamescowens** (Member): "@div72 has a good point."
- **iFoggz** (Member): Acknowledged this should be part of RPC porting from Bitcoin

---

## Root Cause Analysis

### Initial Hypothesis (Incorrect)
Initially suspected the RPC commands were filtering out unconfirmed transactions.

### Actual Root Cause (Confirmed)

**The RPC commands are NOT the problem.** Both Gridcoin and Bitcoin use identical logic:

**Gridcoin** (`src/wallet/rpcwallet.cpp`, line ~2483):
```cpp
// List: Received
if (listReceived.size() > 0 && wtx.GetDepthInMainChain() >= nMinDepth)
```

**Bitcoin** (`src/wallet/rpc/transactions.cpp`, line ~330):
```cpp
// Received
if (listReceived.size() > 0 && wallet.GetTxDepthInMainChain(wtx) >= nMinDepth)
```

Both check `GetDepthInMainChain() >= nMinDepth`, and both pass `nMinDepth = 0`, so transactions with depth 0 (unconfirmed) should appear.

### The Real Problem: Wallet Transaction Tracking

**Bitcoin Core's Architecture**:
```
Mempool Acceptance → SyncTransaction(TxStateInMempool) → Add to mapWallet
Block Confirmation → SyncTransaction(TxStateConfirmed) → Update in mapWallet
Transaction Removal → SyncTransaction(TxStateInactive) → Update in mapWallet
```

**Gridcoin's Current Architecture**:
```
Mempool Acceptance → (no wallet update for incoming transactions)
Block Confirmation → AddToWallet → Add to mapWallet
Transaction Removal → (no specific handling)
```

**Key Difference**:
- Bitcoin: Incoming transactions added to `mapWallet` when they enter mempool
- Gridcoin: Incoming transactions only added to `mapWallet` when confirmed in block

When `listtransactions` iterates through `mapWallet`, incoming unconfirmed transactions simply don't exist yet in Gridcoin's wallet.

### Why Outgoing Transactions Work

Outgoing transactions appear immediately because:
1. Wallet creates the transaction
2. Immediately adds to `mapWallet` (before broadcasting)
3. Transaction is "IsFromMe()" so it's tracked from creation

Incoming transactions don't have this pre-addition step.

---

## Architecture Comparison

### Bitcoin Core Transaction Lifecycle

```
┌─────────────────────────────────────────────────────────┐
│                Transaction State Machine                 │
├─────────────────────────────────────────────────────────┤
│                                                           │
│  [Created/Received]                                      │
│         │                                                │
│         ├──→ TxStateInMempool (0 confirmations)         │
│         │         │                                      │
│         │         └──→ TxStateConfirmed (1+ confirms)   │
│         │         │         │                            │
│         │         │         └──→ (More confirms...)      │
│         │         │                                      │
│         │         └──→ TxStateInactive (conflicted)     │
│         │                                                │
│         └──→ TxStateUnrecognized (unknown)              │
│                                                           │
└─────────────────────────────────────────────────────────┘

Managed by: SyncTransaction(tx, TxState) entry point
Called from: Validation layer callbacks
```

### Gridcoin Current Transaction Lifecycle

```
┌─────────────────────────────────────────────────────────┐
│           Current Transaction Tracking                   │
├─────────────────────────────────────────────────────────┤
│                                                           │
│  [Created/Received]                                      │
│         │                                                │
│         ├──→ (If outgoing: added to mapWallet)          │
│         │                                                │
│         └──→ (If incoming: NOT added until confirmed)   │
│                    │                                     │
│                    └──→ Confirmed in Block               │
│                              │                           │
│                              └──→ AddToWallet            │
│                                        │                 │
│                                        └──→ In mapWallet │
│                                                           │
└─────────────────────────────────────────────────────────┘

No explicit state tracking
No mempool → wallet synchronization for incoming tx
```

### GetDepthInMainChain Comparison

**Bitcoin Core**:
```cpp
int GetTxDepthInMainChain(const CWalletTx& wtx) const {
    if (auto* conf = wtx.state<TxStateConfirmed>()) {
        // Check block is in main chain, return depth
        return nBestHeight - block_height + 1;
    }
    if (wtx.isInMempool() && mempool.exists(hash)) {
        return 0;  // In mempool, unconfirmed
    }
    return -1;  // Not in chain, not in mempool (conflicted)
}
```

**Gridcoin Current**:
```cpp
int CMerkleTx::GetDepthInMainChain(CBlockIndex* &pindexRet) const {
    AssertLockHeld(cs_main);
    int nResult = GetDepthInMainChainINTERNAL(pindexRet);
    if (nResult == 0 && !mempool.exists(GetHash()))
        return -1;  // Not in blockchain AND not in mempool
    return nResult;
}
```

The logic is similar, but it doesn't matter because incoming transactions aren't in `mapWallet` to check!

---

## Proposed Solution

### Solution: Port Bitcoin Core's SyncTransaction Architecture

This involves a comprehensive refactor of wallet transaction tracking to match Bitcoin Core's architecture.

### Core Components to Port

1. **Transaction State Types**
   - `TxStateInMempool` - Transaction in mempool, unconfirmed
   - `TxStateConfirmed` - Transaction in block with height/position
   - `TxStateInactive` - Transaction conflicted or abandoned
   - `TxStateUnrecognized` - Unknown state

2. **SyncTransaction Mechanism**
   - Unified entry point for all transaction updates
   - Called by validation layer on state changes
   - Updates wallet transaction state consistently

3. **Validation Callbacks**
   - `transactionAddedToMempool()`
   - `blockConnected()`
   - `transactionRemovedFromMempool()`
   - `blockDisconnected()`

4. **Updated Transaction Handling**
   - Replace all `AddToWallet()` calls
   - Use state-based transaction tracking
   - Proper state transitions

### Why Option B (Full Port) Over Option A (Quick Fix)

**Option A (Quick Fix)**: Modify existing code to add mempool transactions
- ✅ Faster (2-3 weeks)
- ❌ Doesn't address architectural issues
- ❌ Will need refactoring later anyway
- ❌ Doesn't benefit from Bitcoin's proven architecture

**Option B (Full Port)**: Port Bitcoin's SyncTransaction system
- ✅ Proper architecture (foundation for future)
- ✅ Matches Bitcoin Core (easier to port future improvements)
- ✅ Better state management
- ✅ Cleaner codebase
- ❌ Takes longer (10-12 weeks)

**Decision**: Option B chosen for long-term benefits

---

## Implementation Strategy

### Breaking Into PRs

The implementation will be split into 4 separate Pull Requests:

```
PR1: Transaction State Infrastructure (2 weeks)
  ↓
PR2: SyncTransaction Implementation (2 weeks)
  ↓
PR3: Validation Integration (2 weeks)
  ↓
PR4: Testing & Documentation (2 weeks)
```

(Documents here = https://github.com/PrestackI/clinerules_gridcoin)

### PR1: Transaction State Infrastructure
**Document**: `Issue_1157_PR1_Transaction_States.md`

- Create transaction state types
- Update CWalletTx with state member
- Modify GetDepthInMainChain to use states
- Add state helper methods
- Update serialization

**Deliverables**:
- New file: `src/wallet/transaction.h`
- Modified: `src/wallet/wallet.h`
- Modified: `src/main.cpp` (GetDepthInMainChain)
- Unit tests for state types

### PR2: SyncTransaction Implementation
**Document**: `Issue_1157_PR2_SyncTransaction.md`

- Implement core SyncTransaction function
- Create AddToWalletIfInvolvingMe
- Implement state transition logic
- Add validation callbacks (stubs)
- Transaction lifecycle management

**Deliverables**:
- Modified: `src/wallet/wallet.cpp`
- Modified: `src/wallet/wallet.h`
- Unit tests for SyncTransaction

### PR3: Validation Integration
**Document**: `Issue_1157_PR3_Validation_Integration.md`

- Wire up validation callbacks
- Replace all AddToWallet calls with SyncTransaction
- Integrate with mempool acceptance
- Integrate with block connection/disconnection
- Handle reorgs properly

**Deliverables**:
- Modified: `src/main.cpp`
- Modified: `src/wallet/wallet.cpp`
- Integration tests

### PR4: Testing & Documentation
**Document**: `Issue_1157_PR4_Testing_Documentation.md`

- Comprehensive test suite
- RPC integration tests
- Manual testing procedures
- Documentation updates
- Release notes

**Deliverables**:
- Test files: `src/test/wallet_sync_tests.cpp`
- Documentation: Architecture diagrams
- Updated: RPC command documentation
- Release notes

---

## Timeline & Resources

### Overall Timeline

```
Week 1-2:   PR1 - Transaction State Infrastructure
Week 3-4:   PR2 - SyncTransaction Implementation
Week 5-6:   PR3 - Validation Integration
Week 7-8:   PR4 - Testing & Documentation
Week 9-10:  Code review, fixes, polish
Week 11-12: Buffer for unexpected issues

Total: 10-12 weeks (2.5-3 months)
```

### Resource Requirements

**Developer Skills**:
- Advanced C++ (required)
- Bitcoin/Gridcoin architecture knowledge (required)
- Wallet internals understanding (required)
- Testing experience (required)
- Documentation skills (helpful)

**Time Commitment**:
- Full-time: 10-12 weeks
- Part-time (50%): 20-24 weeks
- Part-time (25%): 40-48 weeks

**Can Be Parallelized**:
- Documentation writing (during implementation)
- Test creation (during development)
- Code review preparation (ongoing)

### Effort Breakdown by PR

| PR | Complexity | Estimated Days | Risk Level |
|----|-----------|----------------|------------|
| PR1 | Medium | 10 days | Low |
| PR2 | High | 10 days | Medium |
| PR3 | High | 10 days | High |
| PR4 | Medium | 10 days | Low |
| Review & Polish | Medium | 10 days | Low |
| Buffer | - | 10 days | - |
| **Total** | - | **60 days** | - |

---

## Success Criteria

### Functional Requirements

✅ **Primary Goal**: Incoming unconfirmed transactions visible in RPC
- `listtransactions` shows incoming tx with 0 confirmations
- `listsinceblock` shows incoming tx with 0 confirmations
- `gettransaction` works with unconfirmed incoming tx

✅ **State Management**: Proper transaction lifecycle
- Mempool → Confirmed transitions work correctly
- Confirmed → Mempool transitions on reorg
- Conflicted/abandoned transactions marked properly

✅ **Backward Compatibility**: No breaking changes
- Existing wallets load without issues
- Old wallet.dat files migrate correctly
- RPC API behavior change only (not signature)

✅ **No Regressions**: Existing functionality preserved
- Outgoing transactions still work
- Balance calculations correct
- Transaction history accurate

### Non-Functional Requirements

✅ **Performance**: No significant degradation
- Wallet operations remain fast
- Synchronization speed maintained
- Memory usage reasonable

✅ **Code Quality**: Professional implementation
- Follows Gridcoin coding standards
- Well-documented code
- Comprehensive test coverage (>80%)

✅ **Maintainability**: Easy to understand and modify
- Clear architecture documentation
- Useful code comments
- Logical organization

### Testing Requirements

✅ **Unit Tests**: Component-level testing
- Transaction state operations
- SyncTransaction logic
- State transitions
- Edge cases

✅ **Integration Tests**: System-level testing
- RPC command behavior
- Wallet synchronization
- Reorg handling
- Multi-transaction scenarios

✅ **Manual Tests**: Real-world validation
- Testnet deployment
- User acceptance testing
- Performance profiling
- Stress testing

---

## Related Issues

### Directly Related

- **Issue #1157**: This issue (incoming unconfirmed transactions)
- **Issue #1142**: Port RPC functionality from Bitcoin (parent issue)

### Indirectly Related

- **Issue #1705** & **PR #1995**: Created `src/rpc/` directory structure
- RPC Porting Analysis Report: Documented gaps in Gridcoin's RPC

### Future Work Enabled

Once this foundation is in place, these become easier:
- Replace-by-fee (RBF) support
- Transaction conflict handling improvements
- Better mempool management
- Advanced wallet features
- PSBT support (future consideration)

---

## Risk Analysis

### Technical Risks

| Risk | Likelihood | Impact | Mitigation |
|------|-----------|--------|------------|
| Breaking existing wallet | Low | Very High | Comprehensive testing, backup procedures |
| Performance degradation | Medium | Medium | Profiling, optimization phase |
| Reorg handling bugs | Medium | High | Extensive reorg testing |
| Migration issues | Low | High | Backward compatibility testing |
| Incomplete state coverage | Medium | Medium | Code review, edge case testing |

### Schedule Risks

| Risk | Likelihood | Impact | Mitigation |
|------|-----------|--------|------------|
| Underestimated complexity | Medium | High | 20% time buffer included |
| Developer availability | Medium | High | Clear documentation for handoff |
| Review bottleneck | Low | Medium | Small PRs, clear docs |
| Testing takes longer | Medium | Medium | Start testing early |

### Mitigation Strategies

1. **Incremental PRs**: Each PR is independently useful
2. **Comprehensive Documentation**: Each PR fully documented
3. **Extensive Testing**: Test at every level
4. **Code Review**: Multiple reviewers for critical changes
5. **Rollback Plan**: Each PR can be reverted if needed
6. **Community Communication**: Keep stakeholders informed

---

## Rollback Strategy

If critical issues are discovered, each PR can be reverted in reverse order:

1. **Revert PR4**: Remove tests/docs (safe, no code impact)
2. **Revert PR3**: Disconnect validation callbacks (wallet functions remain)
3. **Revert PR2**: Remove SyncTransaction (back to old AddToWallet)
4. **Revert PR1**: Remove state infrastructure (back to original)

Each revert returns to a functional state.

---

## Communication Plan

### Stakeholders

- **Users**: Will see unconfirmed incoming transactions
- **Developers**: Need to understand new architecture
- **Exchanges**: May need to update integration code
- **Pool Operators**: May benefit from better transaction visibility

### Documentation Deliverables

1. User-facing: Release notes explaining behavior change
2. Developer-facing: Architecture documentation
3. RPC documentation: Updated command descriptions
4. Migration guide: For exchanges/integrators

### Announcement Timing

- **PR1 Merged**: Internal developer announcement
- **PR2 Merged**: Progress update to community
- **PR3 Merged**: Beta testing call
- **PR4 Merged**: Release candidate announcement
- **Release**: Full release notes and migration guide

---

## Conclusion

This is a significant but necessary architectural improvement that will:
- ✅ Fix the reported issue (unconfirmed transactions visibility)
- ✅ Improve Bitcoin Core compatibility
- ✅ Provide better foundation for future wallet features
- ✅ Align Gridcoin's wallet architecture with modern standards

The incremental PR approach minimizes risk while delivering value progressively.

---

## Next Steps

1. Review and approve this overall plan
2. Review individual PR documents: (Documents here = https://github.com/PrestackI/clinerules_gridcoin)
   - `Issue_1157_PR1_Transaction_States.md`
   - `Issue_1157_PR2_SyncTransaction.md`
   - `Issue_1157_PR3_Validation_Integration.md`
   - `Issue_1157_PR4_Testing_Documentation.md`
3. Begin implementation of PR1
4. Follow Baby Steps™ methodology throughout

---

**Document Version**: 1.0  
**Created**: December 28, 2025  
**Author**: Gridcoin Development Team  
**Status**: Planning Complete, Ready for Implementation
